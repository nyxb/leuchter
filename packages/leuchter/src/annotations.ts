import { annotateLine } from './annotations.inline'
import { annotateLines } from './annotations.multiline'
import type { Annotation } from './comments'
import type { InlineRange, MultiLineRange } from './range'

export interface Token {
   content: string
   style: {
      color?: string
      fontStyle?: 'italic'
      fontWeight?: 'bold'
      textDecoration?: 'underline' | 'line-through'
   }
   scopes?: string[]
}

export interface TokenGroup {
   annotationName: string
   annotationQuery?: string
   fromColumn: number
   toColumn: number
   tokens: Tokens
}

export type Tokens = (Token | TokenGroup)[]

export interface Line { lineNumber: number; tokens: Tokens }
export interface LineGroup {
   annotationName: string
   annotationQuery?: string
   fromLineNumber: number
   toLineNumber: number
   lines: Lines
}
export type Lines = (Line | LineGroup)[]

export function applyAnnotations(lines: Token[][], annotations: Annotation[]) {
   const { inlineAnnotations, multilineAnnotations }
    = splitAnnotations(annotations)

   const annotatedLines = lines.map((line, lineIndex) => {
      const lineAnnotations = inlineAnnotations.filter(
         annotation => annotation.range.lineNumber === lineIndex + 1,
      )
      return {
         lineNumber: lineIndex + 1,
         tokens: annotateLine(line, lineAnnotations),
      } as Line
   })

   return annotateLines(annotatedLines, multilineAnnotations)
}

export type InlineAnnotation = Omit<Annotation, 'ranges'> & {
   range: InlineRange
}

export type MultilineAnnotation = Omit<Annotation, 'ranges'> & {
   range: MultiLineRange
}

function splitAnnotations(annotations: Annotation[]) {
   const inlineAnnotations: InlineAnnotation[] = []
   const multilineAnnotations: MultilineAnnotation[] = []

   annotations.forEach((annotation) => {
      annotation.ranges.forEach((range) => {
         if ('lineNumber' in range)
            inlineAnnotations.push({ ...annotation, range })

         else
            multilineAnnotations.push({ ...annotation, range })
      })
   })

   return {
      inlineAnnotations,
      multilineAnnotations,
   }
}
