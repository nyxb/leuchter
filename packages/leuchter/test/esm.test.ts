import { extractAnnotations, highlight } from '../dist/index.esm.mjs'
import { runAnnotationTests } from './annotations.js'
import { runTests } from './highlight.js'

runTests({ highlight })
runAnnotationTests({ extractAnnotations, highlight })
