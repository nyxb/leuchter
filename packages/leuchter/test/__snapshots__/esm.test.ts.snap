// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extract annottations 1`] = `
{
  "annotations": [
    {
      "name": "foo",
      "query": "",
      "ranges": [
        {
          "fromLineNumber": 2,
          "toLineNumber": 2,
        },
      ],
    },
  ],
  "code": "const x = 1;
const y = 2;",
}
`;

exports[`extract annottations 2`] = `
{
  "lang": "javascript",
  "lines": [
    {
      "lineNumber": 1,
      "tokens": [
        {
          "content": "const ",
          "style": {
            "color": "#569CD6",
          },
        },
        {
          "content": "x",
          "style": {
            "color": "#4FC1FF",
          },
        },
        {
          "content": " = ",
          "style": {
            "color": "#D4D4D4",
          },
        },
        {
          "content": "1",
          "style": {
            "color": "#B5CEA8",
          },
        },
        {
          "content": ";",
          "style": {
            "color": "#D4D4D4",
          },
        },
      ],
    },
    {
      "annotationName": "foo",
      "annotationQuery": "",
      "fromLineNumber": 2,
      "lines": [
        {
          "lineNumber": 2,
          "tokens": [
            {
              "content": "const ",
              "style": {
                "color": "#569CD6",
              },
            },
            {
              "content": "y",
              "style": {
                "color": "#4FC1FF",
              },
            },
            {
              "content": " = ",
              "style": {
                "color": "#D4D4D4",
              },
            },
            {
              "content": "2",
              "style": {
                "color": "#B5CEA8",
              },
            },
            {
              "content": ";",
              "style": {
                "color": "#D4D4D4",
              },
            },
          ],
        },
      ],
      "toLineNumber": 2,
    },
  ],
  "style": {
    "background": "#1E1E1E",
    "color": "#D4D4D4",
  },
}
`;

exports[`extract annottations from text 1`] = `
{
  "annotations": [],
  "code": "const x = 1;
// foo[3:5]
const y = 2;",
}
`;

exports[`extract annottations from text 2`] = `
{
  "lang": "text",
  "lines": [
    {
      "lineNumber": 1,
      "tokens": [
        {
          "content": "const x = 1;",
          "style": {},
        },
      ],
    },
    {
      "lineNumber": 2,
      "tokens": [
        {
          "content": "// foo[3:5]",
          "style": {},
        },
      ],
    },
    {
      "lineNumber": 3,
      "tokens": [
        {
          "content": "const y = 2;",
          "style": {},
        },
      ],
    },
  ],
  "style": {
    "background": "#1E1E1E",
    "color": "#D4D4D4",
  },
}
`;

exports[`highlight html with theme 1`] = `
{
  "lang": "html",
  "lines": [
    [
      {
        "content": "<",
        "style": {
          "color": "#C9D1D9",
        },
      },
      {
        "content": "script",
        "style": {
          "color": "#7EE787",
        },
      },
      {
        "content": "> ",
        "style": {
          "color": "#C9D1D9",
        },
      },
      {
        "content": "const ",
        "style": {
          "color": "#FF7B72",
        },
      },
      {
        "content": "x ",
        "style": {
          "color": "#79C0FF",
        },
      },
      {
        "content": "= ",
        "style": {
          "color": "#FF7B72",
        },
      },
      {
        "content": "1 ",
        "style": {
          "color": "#79C0FF",
        },
      },
      {
        "content": "</",
        "style": {
          "color": "#FF7B72",
        },
      },
      {
        "content": "script",
        "style": {
          "color": "#C9D1D9",
        },
      },
      {
        "content": ">",
        "style": {
          "color": "#FF7B72",
        },
      },
    ],
  ],
  "style": {
    "background": "#0d1117",
    "color": "#c9d1d9",
  },
}
`;

exports[`highlight js 1`] = `
{
  "lang": "javascript",
  "lines": [
    [
      {
        "content": "x",
        "style": {
          "color": "#9CDCFE",
        },
      },
      {
        "content": " = ",
        "style": {
          "color": "#D4D4D4",
        },
      },
      {
        "content": "1",
        "style": {
          "color": "#B5CEA8",
        },
      },
    ],
  ],
  "style": {
    "background": "#1E1E1E",
    "color": "#D4D4D4",
  },
}
`;

exports[`highlight text 1`] = `
{
  "lang": "text",
  "lines": [
    [
      {
        "content": "x = 1",
        "style": {},
      },
    ],
  ],
  "style": {
    "background": "#1E1E1E",
    "color": "#D4D4D4",
  },
}
`;

exports[`highlight with empty theme 1`] = `
{
  "lang": "javascript",
  "lines": [
    [
      {
        "content": "x = 1",
        "style": {
          "color": "#BBBBBB",
        },
      },
    ],
  ],
  "style": {
    "background": "#1E1E1E",
    "color": "#bbbbbb",
  },
}
`;

exports[`highlight with frozen theme 1`] = `
{
  "lang": "javascript",
  "lines": [
    [
      {
        "content": "x = 1",
        "style": {
          "color": "#BBBBBB",
        },
      },
    ],
  ],
  "style": {
    "background": "#1E1E1E",
    "color": "#bbbbbb",
  },
}
`;

exports[`highlight with frozen theme with global setting 1`] = `
{
  "lang": "javascript",
  "lines": [
    [
      {
        "content": "x = 1",
        "style": {
          "color": "#BBBBBB",
        },
      },
    ],
  ],
  "style": {
    "background": "#1E1E1E",
    "color": "#bbbbbb",
  },
}
`;

exports[`highlight with scopes 1`] = `
{
  "lang": "javascript",
  "lines": [
    [
      {
        "content": "x",
        "scopes": [
          "variable.other.readwrite.js",
          "source.js",
        ],
        "style": {
          "color": "#C9D1D9",
        },
      },
      {
        "content": " ",
        "scopes": [
          "source.js",
        ],
        "style": {
          "color": "#C9D1D9",
        },
      },
      {
        "content": "=",
        "scopes": [
          "keyword.operator.assignment.js",
          "source.js",
        ],
        "style": {
          "color": "#FF7B72",
        },
      },
      {
        "content": " ",
        "scopes": [
          "source.js",
        ],
        "style": {
          "color": "#C9D1D9",
        },
      },
      {
        "content": "1",
        "scopes": [
          "constant.numeric.decimal.js",
          "source.js",
        ],
        "style": {
          "color": "#79C0FF",
        },
      },
    ],
  ],
  "style": {
    "background": "#0d1117",
    "color": "#c9d1d9",
  },
}
`;

exports[`null code 1`] = `
{
  "lang": "javascript",
  "lines": [
    [
      {
        "content": "",
        "style": {
          "color": "#D4D4D4",
        },
      },
    ],
  ],
  "style": {
    "background": "#1E1E1E",
    "color": "#D4D4D4",
  },
}
`;

exports[`null lang 1`] = `
{
  "lang": "text",
  "lines": [
    [
      {
        "content": "x = 1",
        "style": {},
      },
    ],
  ],
  "style": {
    "background": "#1E1E1E",
    "color": "#D4D4D4",
  },
}
`;
